src/VisionModel/resnet.jl

```diff
@@ -0,0 +1,93 @@
+using Flux
+
+export ResNet18, ResNet34, ResNet50, ResNet101, ResNet152
+
+basicblock(inplanes, outplanes, downsample = false) = downsample ? 
+  Chain(Conv((3, 3), inplanes => outplanes[1], stride = 2),
+        BatchNorm(outplanes[1], relu),
+        Conv((3, 3), outplanes[1] => outplanes[2], stride = 1, pad = 1),
+        BatchNorm(outplanes[2], relu)) : 
+  Chain(Conv((3, 3), inplanes => outplanes[1], stride = 1),
+        BatchNorm(outplanes[1], relu),
+        Conv((3, 3), outplanes[1] => outplanes[2], stride = 1, pad = 1),
+        BatchNorm(outplanes[2], relu))
+
+bottleneck(inplanes, outplanes, downsample = false) = downsample ?
+  Chain(Conv((1, 1), inplanes => outplanes[1], stride = 2),
+        BatchNorm(outplanes[1], relu),
+        Conv((3, 3), outplanes[1] => outplanes[2], stride = 1, pad = 1),
+        BatchNorm(outplanes[2], relu),
+        Conv((1, 1), outplanes[2] => outplanes[3], stride = 1),
+        BatchNorm(outplanes[3], relu)) :
+  Chain(Conv((1, 1), inplanes => outplanes[1], stride = 1),
+        BatchNorm(outplanes[1], relu),
+        Conv((3, 3), outplanes[1] => outplanes[2], stride = 1, pad = 1),
+        BatchNorm(outplanes[2], relu),
+        Conv((1, 1), outplanes[2] => outplanes[3], stride = 1),
+        BatchNorm(outplanes[3], relu))
+
+function projection(inplanes, outplanes, stride)
+  shortcut = Chain(Conv((1, 1), inplanes => outplanes, stride=stride),
+                   BatchNorm((outplanes), relu))
+  return (x, y) -> x + shortcut(y)
+end
+
+# array -> PaddedView(0, array, outplanes) for zero padding arrays
+function identity(inplanes, outplanes, stride)
+  shortcut = (outplanes[1] > inplanes) ? 
+    array ->  cat(array, zeros(eltype(array), outplanes[1] - inplanes); dims = (1, 2, 3)) : +
+  return (x, y) -> x + shortcut(y)
```

Comment:
So this isn't exactly what we want. Right now, when `outplanes[1] <= inplanes`, `shortcut = +` which is a binary function. This won't work with L39. The end result that we want is if `outplanes[1] > inplanes`, then the returned function is `(x, y) -> x + shortcut(y)` (where `shortcut = array -> cat(...)`). If `outplanes[1] <= inplanes`, then the returned functions is `+`. You could do this like
```julia
function identity(inplanes, outplanes, stride)
    shortcut = array ->  cat(array, zeros(eltype(array), outplanes[1] - inplanes); dims = (1, 2, 3))
    return (outplanes[1] > inplanes) ? (x, y) -> x + shortcut(y) : +
end
```

Reply:
That's not working either

Reply:
Yeah, while that fix is necessary, I don't think it is responsible for the error you are seeing. I would commit and using print statements to debug. I'm not saying that the model building is totally wrong, but we need to debug to find out what's causing the error. Since the code was never tested properly before, we can't be certain that its these specific changes causing the error. The error could have been there all along, and we didn't discover it until now.

Reply:
The dimension mismatch is due to ``+`` in ``identity`` function.

Reply:
But it could also be because the basic block is not being passed the right parameters, or even with the right parameters the logic is wrong, and thus, the output of the basic block is not the right shape. That issue will appear when you try to do `x + y` or `x + shortcut(y)`, but the source of the issue may be further back in the code.

The cat operation should just be adding extra channels. And since the channels matched up in the error, I don't think the cat operation is wrong. It is possible that the cat operation is not doing what I think it should do, and it is incorrectly adjusting the width and height dimensions, but you can easily test this for a random array in the REPL.

The `+` operation is very simple. If it is throwing an error because the sizes don't match, then it is from an earlier line of code. `+` doesn't modify the size of its inputs.

Assuming the REPL test for cat works, I would say the best thing to do is print out the input size to every layer and see if they make sense with what you expect by working through the network by hand. I believe the error is within the first few layers so it should be quick to catch this way. Can you commit your latest code so that we can both attempt to debug?

Reply:
I haven't changed anything since last commit except adding few println statements

Reply:
Let me check the cat function in REPL

Reply:
I don't think your latest commit includes the fix at the start of this thread

Reply:
Yeah, sorry. 
I have changed ``outplanes[1]`` to ``outplanes[end]`` and also one identity shortcut is calling with ``outplanes`` while other is calling with ``outplanes[end]``. 
```
    if shortcut_config == :A
      push!(layers, SkipConnection(block(inplanes, outplanes, true),
                                   identity(inplanes, outplanes, 2))) ---------------------(1)
    elseif shortcut_config == :B || shortcut_config == :C
      push!(layers, SkipConnection(block(inplanes, outplanes, true),
                                   projection(inplanes, outplanes[end], 2)))
    end
    inplanes = outplanes[end]
    for i in 2:nrepeats
      if shortcut_config == :A || shortcut_config == :B
        push!(layers, SkipConnection(block(inplanes, outplanes, false),
                                     identity(inplanes, outplanes[end], 1))) -----------------(2)
```

Reply:
The basicblock has certainly some issues bt I am not able to find those out. Here is the stack trace:
```
DimensionMismatch("dimensions must match: a has dims (Base.OneTo(31), Base.OneTo(31), Base.OneTo(64), Base.OneTo(50)), b has dims (Base.OneTo(64), Base.OneTo(64), Base.OneTo(64), Base.OneTo(50)), mismatch at 1")
  Stacktrace:
   [1] promote_shape at ./indices.jl:178 [inlined]
   [2] promote_shape(::Array{Float32,4}, ::Array{Float32,4}) at ./indices.jl:169
   [3] +(::Array{Float32,4}, ::Array{Float32,4}) at ./arraymath.jl:45
   [4] (::SkipConnection)(::Array{Float32,4}) at /home/nabanita07/.julia/packages/Flux/IjMZL/src/layers/basic.jl:259
   [5] applychain(::Tuple{SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,AdaptiveMeanPool{4,2},typeof(flatten),Dense{typeof(identity),Array{Float32,2},Array{Float32,1}}}, ::Array{Float32,4}) at /home/nabanita07/.julia/packages/Flux/IjMZL/src/layers/basic.jl:36 (repeats 4 times)
   [6] (::Chain{Tuple{Conv{2,2,typeof(identity),Array{Float32,4},Array{Float32,1}},BatchNorm{typeof(relu),Array{Float32,1},Array{Float32,1},Float32},MaxPool{2,2},SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,AdaptiveMeanPool{4,2},typeof(flatten),Dense{typeof(identity),Array{Float32,2},Array{Float32,1}}}})(::Array{Float32,4}) at /home/nabanita07/.julia/packages/Flux/IjMZL/src/layers/basic.jl:38
   [7] macro expansion at /home/nabanita07/FluxModels.jl/test/VisionModel/resnet.jl:8 [inlined]
   [8] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
   [9] top-level scope at /home/nabanita07/FluxModels.jl/test/VisionModel/resnet.jl:6
  
```